wallet/src/restore.rs

```diff
@@ -103,85 +112,158 @@ pub fn utxos_batch_block(config: &WalletConfig, start_height: u64, end_height:u6
-				break;
 			}
 		}
 	}
+	debug!(
+		LOGGER,
+		"Found {} wallet_outputs for block {}",
+		wallet_outputs.len(),
+		block_outputs.header.height,
+	);
+
 	wallet_outputs
 }
 
-pub fn restore(config: &WalletConfig, keychain: &Keychain, key_derivations:u32) ->
-	Result<(), Error>{
+pub fn restore(
+	config: &WalletConfig,
+	keychain: &Keychain,
+	key_derivations: u32,
+) -> Result<(), Error> {
 	// Don't proceed if wallet.dat has anything in it
 	let is_empty = WalletData::read_wallet(&config.data_file_dir, |wallet_data| {
-		wallet_data.outputs.len() ==  0
+		wallet_data.outputs.len() == 0
 	})?;
 	if !is_empty {
-		error!(LOGGER, "Not restoring. Please back up and remove existing wallet.dat first.");
-		return Ok(())
+		error!(
+			LOGGER,
+			"Not restoring. Please back up and remove existing wallet.dat first."
+		);
+		return Ok(());
 	}
 
-// Get height of chain from node (we'll check again when done)
+	// Get height of chain from node (we'll check again when done)
 	let chain_height = get_chain_height(config)?;
-	info!(LOGGER, "Starting restore: Chain height is {}.", chain_height);
+	info!(
+		LOGGER,
+		"Starting restore: Chain height is {}.",
+		chain_height
+	);
 
-	let mut switch_commit_cache : Vec<[u8;SWITCH_COMMIT_HASH_SIZE]> = vec![];
-	info!(LOGGER, "Building key derivation cache to index {}.", key_derivations);
+	let mut switch_commit_cache: Vec<[u8; SWITCH_COMMIT_HASH_SIZE]> = vec![];
+	info!(
+		LOGGER,
+		"Building key derivation cache ({}) ...",
+		key_derivations,
+	);
 	for i in 0..key_derivations {
 		let switch_commit = keychain.switch_commit_from_index(i as u32).unwrap();
 		let switch_commit_hash = SwitchCommitHash::from_switch_commit(switch_commit);
 		switch_commit_cache.push(switch_commit_hash.hash);
 	}
+	debug!(LOGGER, "... done");
 
-	let batch_size=100;
-	//this will start here, then lower as outputs are found, moving backwards on the chain
-	let mut key_iterations=key_derivations as usize;
-	//set to a percentage of the key_derivation value
-	let mut padding = (key_iterations as f64 *0.25) as usize;
+	let batch_size = 100;
+	// this will start here, then lower as outputs are found, moving backwards on
+	// the chain
+	let mut key_iterations = key_derivations as usize;
+	// set to a percentage of the key_derivation value
+	let mut padding = (key_iterations as f64 * 0.25) as usize;
 	let mut h = chain_height;
 	while {
-		let end_batch=h;
+		let end_batch = h;
 		if h >= batch_size {
-			h-=batch_size;
+			h -= batch_size;
 		} else {
-			h=0;
+			h = 0;
 		}
-		let mut blocks = utxos_batch_block(config, h+1, end_batch)?;
+		let mut blocks = utxos_batch_block(config, h + 1, end_batch)?;
 		blocks.reverse();
-		for block in blocks {
-			let result_vec=find_utxos_with_key(config, keychain, &switch_commit_cache,
-				block, &mut key_iterations, &mut padding);
-			if result_vec.len() > 0 {
-				for output in result_vec.clone() {
-					let _ = WalletData::with_wallet(&config.data_file_dir, |wallet_data| {
```

Comment:
Moved the wallet lock out of the loop (now we get it once per batch of blocks).
We need to be running `wallet restore` without any other processes touching `wallet.dat` anyway.